src/api.cu

```diff
@@ -165,130 +168,163 @@ TRITONBACKEND_ModelInstanceExecute(
+      raw_requests, raw_requests + request_count);
+
+  auto instance_state = get_instance_state<ModelInstanceState>(*instance);
+  auto model_state = instance_state->StateForModel();
+  auto target_memory = get_native_memory_for_instance(instance_state->Kind());
+
+  auto construct_input_batches = [&](){
+    return get_input_batches<float>(
+      uint32_t{0}, requests, target_memory,
+      instance_state->get_raft_handle());
+  };
+
+  // Get all input data and bail out gracefully if this fails
+  decltype(construct_input_batches()) input_batches;
   try {
-    auto instance_state = get_instance_state<ModelInstanceState>(*instance);
-    ModelState* model_state = instance_state->StateForModel();
-    auto target_memory = get_native_memory_for_instance(instance_state->Kind());
+    input_batches = construct_input_batches();
+  }
+  catch (TritonException& setup_err) {
+    return setup_err.error();
+  }
 
-    std::vector<TRITONBACKEND_Request*> requests(
-        raw_requests, raw_requests + request_count);
+  // BEGIN: Lambda definitions for processing individual batches
+
+  /* Given an input batch and output batch, perform prediction on input,
+   * storing results in output and returning timing information. Return value
+   * is a std::optional containing a pair of uint64_t representing nanoseconds
+   * since the epoch to the beginning and end of prediction respectively. If
+   * std::optional is std::nullopt, prediction has failed and timing values
+   * would have no meaning */
+  auto instance_predict = [&](auto& input_batch, auto& output_batch) {
+    auto start_time =
+        uint64_t(std::chrono::steady_clock::now().time_since_epoch().count());
+    instance_state->predict(
+        input_batch.data, output_batch, model_state->predict_proba);
+    auto end_time =
+        uint64_t(std::chrono::steady_clock::now().time_since_epoch().count());
+    output_batch.sync();
+    return std::make_pair(start_time, end_time);
+  };
+
+
+  /* Given a vector of input batches, return a vector of vectors representing
+   * the output shape of each batch */
+  auto get_output_shapes = [&](auto& batch) {
+    std::vector<std::vector<int64_t>> output_shapes;
+    output_shapes.reserve(batch.shapes.size());
+    std::transform(
+        batch.shapes.begin(), batch.shapes.end(),
+        std::back_inserter(output_shapes), [&](auto& input_shape) {
+          return model_state->get_output_shape(input_shape);
+        });
+    return output_shapes;
+  };
+
+  /* Given start and end iterators for requests, process them and send
+   * responses. Return a std::optional containing a pair of values representing
+   * the time in nanoseconds since the epoch at which computation began and
+   * ended for this batch.
+   *
+   * This lambda is guaranteed to send a response for each request in the batch
+   */
+  auto respond_to_requests = [&](auto& batch,
+                                 decltype(requests.begin()) requests_begin,
```

Comment:
why no `auto` here?

Reply:
No good reason, as it turns out :smile:. I had convinced myself that the type could not properly be deduced down at line 242, but it totally can... if I had remembered to include a `typename` in there! Fixed it, and now we're looking much more pleasantly `auto`ish.